## Практическое задание по теме "Введение в линейную алгебру"

### Задание 5.1.

Вектор – это частный случай матрицы 1хN и Nх1. Повторите материал для векторов, уделяя особое внимание умножению A∙B.
Вычислите, по возможности не используя программирование: $(5E)^{-1}$ где Е – единичная матрица размера 5х5.

#### Решение
$$
(
5 \cdot
\begin{pmatrix}
1 & 0 & 0 & ... & 0 \\
0 & 1 & 0 & ... & 0 \\
0 & 0 & 1 & ... & 0 \\
... & ... & ... & ... & 0 \\
0 & 0 & 0 & ... & 1 \\
\end{pmatrix}
)^{-1}
=
\begin{pmatrix}
\frac{1}{5} & 0 & 0 & ... & 0 \\
0 & \frac{1}{5} & 0 & ... & 0 \\
0 & 0 & \frac{1}{5} & ... & 0 \\
... & ... & ... & ... & 0 \\
0 & 0 & 0 & ... & \frac{1}{5} \\
\end{pmatrix}
$$

### Задание 5.2

Вычислите определитель:



#### Решение

$
\det{
\begin{bmatrix}
1 & 2 & 3 \\
4 & 0 & 6 \\
7 & 8 & 9
\end{bmatrix}
} =
1 \cdot
\begin{bmatrix}
0 & 6 \\
8 & 9
\end{bmatrix}
- 2 \cdot
\begin{bmatrix}
4 & 6 \\
7 & 9
\end{bmatrix}
+ 3 \cdot
\begin{bmatrix}
4 & 0 \\
7 & 8
\end{bmatrix}
= - 48 - 2(36 - 42) + 96 = 12 + 96 - 48 = 60
$

#### Проверка

In [98]:
import numpy as np
A = np.array([[1, 2, 3], [4, 0, 6], [7, 8, 9]])
np.linalg.det(A)

59.999999999999986

### Задание 5.3

#### 1. Вычислите матрицу, обратную данной:

$$
A=
\begin{vmatrix}
1 & 2 & 3 \\
4 & 0 & 6 \\
7 & 8 & 9
\end{vmatrix}
$$

Рассчитаем матрицу алгебраических дополнений

$$
A_{*}=
\begin{bmatrix}
A_{11} & A_{12} & A_{13} \\
A_{21} & A_{22} & A_{23} \\
A_{31} & A_{32} & A_{33}
\end{bmatrix}
=
\begin{vmatrix}
-48 & -6 & 32 \\
12 & -12 & 6 \\
12 & 6 & -8
\end{vmatrix}
$$

Транспонированная матрица алгебраических дополнений будет иметь следующий вид

$$
A_{*}^T = 
\begin{vmatrix}
-48 & 12 & 12 \\
-6 & -12 & 6 \\
32 & 6 & -8
\end{vmatrix}
$$

Определитель матрицы будет равен

$$
\det{A} = - 48 - 2(36 - 42) + 96 = 12 + 96 - 48 = 60
$$

Тогда обратная матрица будет иметь вид:

$$
A^{-1}= \frac{A_{*}^T}{\det{A}} = 
\frac{1}{60}\cdot
\begin{vmatrix}
-48 & 12 & 12 \\
-6 & -12 & 6 \\
32 & 6 & -8
\end{vmatrix}
=
\begin{vmatrix}
-0.8 & 0.2 & 0.2 \\
-0.1 & -0.2 & 0.1 \\
\frac{8}{15} & 0.1 & -\frac{2}{15}
\end{vmatrix}
$$

#### Проверка

In [99]:
import numpy as np

A = [[1, 2, 3], [4, 0, 6], [7, 8, 9]]
A_inv = np.linalg.inv(A)
A_inv

array([[-0.8       ,  0.1       ,  0.2       ],
       [ 0.1       , -0.2       ,  0.1       ],
       [ 0.53333333,  0.1       , -0.13333333]])

#### 2. Приведите пример матрицы 4х4, ранг которой равен 1.

$$
\begin{vmatrix}
0 & 0 & 0 & 1 \\
0 & 0 & 0 & 1 \\
0 & 0 & 0 & 1 \\
0 & 0 & 0 & 1 \\
\end{vmatrix}
$$

In [100]:
import numpy as np
A = [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1]]
np.linalg.matrix_rank(A)

1

### Задание 5.4

Вычислите скалярное произведение двух векторов:  
(1, 5)  и  (2, 8)

#### Решение

$$(1, 5)\cdot(2, 8)= 1\cdot2 + 5\cdot8 = 42$$

### Задание 5.5

Вычислите смешанное произведение трех векторов:
(1, 5, 0), (2, 8, 7) и (7, 1.5, 3) 

#### Решение

$
V =
\begin{vmatrix}
1 & 5 & 0 \\
2 & 8 & 7 \\
7 & 1.5 & 3
\end{vmatrix}
= 24 - 10.5 - 5(6 - 49) = 13.5 - 30 + 245 = 228.5
$

#### Проверка

In [101]:
a = np.array([1, 5, 0], float)
b = np.array([2, 8, 7], float)
c = np.array([7, 1.5, 3], float)

v = np.cross(a, b)
result = np.inner(v, c)
result

228.5

### Задания по теме линейных систем уравнений

### Задание 1

$$
\begin{vmatrix}
1 & 2 & 3 \\
4 & 0 & 6 \\
7 & 8 & 9
\end{vmatrix}
\cdot
X=
\cdot
\begin{vmatrix}
12 \\ 2 \\ 1
\end{vmatrix}
$$

#### Решение

$$
\begin{cases}
x_0 + 2x_1 + 3x_2 = 12 \\
4 x_0 + 6x_2 = 2 \\
7x_0 + 8x_1 + 9x_2 = 1
\end{cases}
\implies
\begin{cases}
x_0 + 2x_1 + 3x_2 = 12 \\
x_2 = \frac{1}{3} - \frac{2}{3}x_0 \\
7x_0 + 8x_1 + 9x_2 = 1
\end{cases}
\implies
\begin{cases}
x_0 + 2x_1 + 1 - 2x_0 = 12 \\
x_2 = \frac{1}{3} - \frac{2}{3}x_0 \\
7x_0 + 8x_1 + 9x_2 = 1
\end{cases}
\implies
\begin{cases}
x_1 = \frac{11 + x_0}{2} \\
x_2 = \frac{1}{3} - \frac{2}{3}x_0 \\
7x_0 + 8x_1 + 9x_2 = 1
\end{cases}
\implies
\begin{cases}
x_1 = \frac{11 + x_0}{2} \\
x_2 = \frac{1}{3} - \frac{2}{3}x_0 \\
7x_0 + 4(11 + x_0) + 3(1 - 2x_0) = 1
\end{cases}
\implies
\begin{cases}
x_1 = \frac{11 + x_0}{2} \\
x_2 = \frac{1}{3} - \frac{2}{3}x_0 \\
x_0 = -9.2
\end{cases}
\implies
\begin{cases}
x_1 = 0.9 \\
x_2 = \frac{1}{3} + \frac{18.4}{3} = 6.47\\
x_0 = -9.2
\end{cases}
$$

#### Проверка

In [102]:
A = np.array([[1, 2, 3], [4, 0, 6], [7, 8, 9]])
B = np.array([12, 2, 1])
np.linalg.solve(A, B)

array([-9.2       ,  0.9       ,  6.46666667])

### Задание 2

#### Найдите псевдорешение:
$$
\begin{cases}
x + 2y – z = 1 \\
3x – 4y = 7 \\
8x – 5y + 2z = 12 \\
2x – 5z = 7 \\
11x +4y – 7z = 15
\end{cases}
$$

#### Решение

In [103]:
A = np.array([[1, 2, -1], [3, -4, 0], [8, -5, 2], [2, -5, 0], [11, 4, -7]])
B = np.array([1, 7, 12, 7, 15])
np.linalg.lstsq(A, B, rcond=None)

(array([ 1.15601029, -0.889718  , -0.85029782]),
 array([0.66772534]),
 3,
 array([14.89650167, 10.4695886 ,  2.73531577]))

Получившееся псевдорешение

$$
C = [1.15601029, -0.889718  , -0.85029782]
$$

### Задание 3

Сколько решений имеет система:

$$
\begin{vmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9
\end{vmatrix}
\cdot
X=
\cdot
\begin{vmatrix}
12 \\ 2 \\ 1
\end{vmatrix}
$$

Если ноль – то измените вектор правой части так, чтобы система стала совместной, и решите ее. 

#### Решение

In [104]:
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
B = np.array([12, 22, 23])

np.linalg.matrix_rank(A)

2

Система недоопределена, т.к. ранк матрицы меньше количества неизвестных

In [107]:
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
B = np.array([0, 0, 0])
np.linalg.solve(A, B)

array([0., 0., 0.])

### Задание 4

Вычислите LU-разложение матрицы:

$$
\begin{vmatrix}
1 & 2 & 3 \\
2 & 16 & 21 \\
4 & 28 & 73 
\end{vmatrix}
$$

После этого придумайте вектор правых частей и решите полученную линейную систему трех уравнений с данной матрицей.

#### Решение

In [114]:
import scipy
import scipy.linalg

A = np.array([[1, 2, 3], [2, 16, 21], [4, 28, 73]])
P, L, U = scipy.linalg.lu(A)
P, L, U

(array([[0., 1., 0.],
        [0., 0., 1.],
        [1., 0., 0.]]), array([[ 1.  ,  0.  ,  0.  ],
        [ 0.25,  1.  ,  0.  ],
        [ 0.5 , -0.4 ,  1.  ]]), array([[  4.  ,  28.  ,  73.  ],
        [  0.  ,  -5.  , -15.25],
        [  0.  ,   0.  , -21.6 ]]))

Составим вектор B, приняв значения переменных $x_0=x_1=x_2=1$

### Задание 5

#### Решение